example/http/client/crawl/http_crawl.cpp

```diff
@@ -356,7 +356,10 @@ int main(int argc, char* argv[])
     net::io_context ioc;
 
     // The work keeps io_context::run from returning
-    auto work = net::make_work_guard(ioc);
+    auto work = net::any_io_executor(
+        net::prefer(
```

Comment:
I have a preference for using `require` instead of `prefer` for getting the `outstanding_work.tracked` property in this particular kind of context. My work-in-progress rule of thumb is:

* Use `require` if all of the following are true for the use case:
  - The executor is a known, concrete type (an `io_context` executor type here) rather than a template parameter
  - The concrete executor is known to support the `outstanding_work.tracked` property
  - The correctness of the code in question (non-generic code; the whole program in this case) needs work to be tracked

* Use `prefer` if the following are all true:
  - The code is generic and the executor type is a template parameter, dependent type, or `any_io_executor`
  - The executor is not known or not guaranteed to support the `outstanding_work.tracked` property
  - The correctness of the code (e.g. a generic, templated algorithm or class, or a runtime-polymorphic algorithm) does not in and of itself depend on work being tracked; i.e. think of it as generic code in a self-contained, testable unit

In the above line, we require work to be tracked against the `io_context` for the program to work, so we use `require`.

Reply:
The old `work_guard` was much easier to use and understand, why are we getting rid of it?

Reply:
It's not officially being got rid of from Asio (yet), but it doesn't really fit the model in the same way any more. This is because `executor_work_guard` is an RAII adapter over a resource acquisition function `Executor.on_work_started()` and a resource release function `Executor.on_work_started()`. With the proposed standard executors, the executor *is* the RAII type.

I took a closer look at this example, and actually I think it might be able to do its work tracking in a different, hopefully clearer way. I've pushed two branches with some experimental refactoring for this:

https://github.com/chriskohlhoff/Beast/commit/3f7ef481251e35f9c33c592958d13df2cbce0062
https://github.com/chriskohlhoff/Beast/commit/b58fd9698b8faf8dbdea767b5d2b9c82e7a33987

There are no doubt more variations on these themes.

I particularly like the first because it reflects the fact that each worker thread represents the promise of some future delivery of aggregation work, and that promise corresponds to an obligation to represent that future delivery as outstanding work. I think this could be a better way of thinking about it in this example even when using `make_work_guard`, and seems less arbitrary than capturing outstanding work simply to stop an `io_context` from running out of things to do.

Reply:
I had a similar notion but I wasn't able to put it in such elegant terms, thanks!

Reply:
@chriskohlhoff To be honest, having now had some real-world use, I am of the view that asio's work_guard object should remain. It's purpose was (and still is) explicit. Whether it is implemented in terms of properties is of no consequence. The fact that it has a singular purpose makes it easy to identify in code and easy to teach.

The reaction we've observed from colleagues when presenting property-based executor manipulation in mainline code is along the lines of "WTF! Why?"

You and I of course know why technically, but in my view it adds very little in terms of expressing intent.